rust/frontend/src/server.rs

```diff
@@ -313,6 +314,10 @@ impl FrontendServer {
         let addr = format!("{}:{}", listen_address, port);
         println!("Listening on {addr}");
         let listener = tokio::net::TcpListener::bind(addr).await.unwrap();
+        let bound_port = listener.local_addr().unwrap().port();
```

Comment:
nit - prefer expect to unwrap. 

Comment with 

// SAFETY(<username>) reason why you think this is safe